### PR TITLE
have FileLoader merge multiple concurrent requests for the same url

### LINKED ISSUE #10644: If cache is enabled, FileLoader should not make duplicate requests.
##### Description of the problem 

Given the code below, only one request should be fired.

```js
THREE.Cache.enabled = true
var fileLoader = new THREE.FileLoader();
fileLoader.load('foo.jpg');  // A
fileLoader.load('foo.jpg');  // B
```

1. `A` checks the cache and sees no pending requests.
2. `A` makes the request and stores the pending request in the cache as a Promise. The Promise will resolve the file.
3. `B` checks the cache and sees a pending request. `B` waits for the Promise to resolve the file.
4. The request finishes, and the Promise resolves.
5. `A` and `B`, both waiting on the Promise, triggers their callbacks at the same time with the same file response.

I'd like to volunteer to work on this. In A-Frame, we often have been having to maintain a cache of promises to prevent duplicate requests. If it sounds reasonable, below are several possible implementations:

1. Change THREE.Cache to store promises that resolve files, rather than directly storing files. I think this is the optimal solution because requests will always be asynchronous. This is sort of a breaking change, but three.js applications don't often enable the cache, or don't care about the internals of the cache.

2. Create a separate cache object (`RequestCache`) for pending requests. The current Cache stays intact, but there is more logic to maintain both the pending and complete caches.

3. Keep the pending request cache internal to the FileLoader. It'll just be a variable `var requestCache = {}`.

We'd also need a Promise polyfill. They're less than 1KB gzipped if that sounds fine.

##### Three.js version

- [x] Dev
- [x] r84

##### Browser

- [x] All of them

##### OS

- [x] All of them

##### Hardware Requirements (graphics card, VR Device, ...)

N/A

### PR DESCRIPTION
**Pull request by ngokevin**
For multiple requests to the same file:

1. The first request makes the request. Creates an array for other loaders to subscribe callbacks to.
2. When other loaders try to make the same request, they append a callback that re-calls their `.load`.
3. When the request finishes, all the callbacks are called.
4. Every loader loads their file from cache.

This wasn't necessary to do for ImageLoader since the browser will handle multiple requests to the same image internally.

Avoided Promises since the rest of the codebase didn't use it, just for old browser compatibility.

## Before

Multiple models of the same URL loading at the same time, multiple requests. Tested at the A-Frame level which has asynchrony/DOM that would cause multiple duplicate requests:

screen shot 2017-10-18 at 4 22 50 pm

## After

Multiple models of the same URL loading at the same time, one request:

screen shot 2017-10-18 at 4 18 42 pm

Fixes #10644

### CODE REVIEW

File: src/loaders/FileLoader.js [hunk 1]
```diff
@@ -39,6 +39,19 @@ Object.assign( FileLoader.prototype, {
 
 		}
 
+		// If file is already in process of loading, wait for it to load.
+		if ( cached && cached.loaderSubscriptions ) {
```
Comment (takahirox) [inline src/loaders/FileLoader.js]: Can we replace `cached` with `cached !== undefined` here for the consistent code in the function?

**Comment by takahirox** [comment]
It isn't serious issue but wanna note just in case that `Image(Bitmap)Loader.load()` loads wrong data(`{ loaderSubscriptions: loaderSubscriptions }`) if it loads right after `FileLoader.load()` loads the same url like this

    var fileLoader = new THREE.FileLoader();
    var imageLoader = new THREE.ImageLoader();
    fileLoader.load( url );
    imageLoader.load( url );

(Probably no one runs like this)

**Comment by ngokevin** [comment]
Oops, I should maintain a separate object versus storing that stuff in cache.

File: src/loaders/FileLoader.js [hunk 2]
```diff
@@ -129,6 +144,14 @@ Object.assign( FileLoader.prototype, {
...
 			var request = new XMLHttpRequest();
+
+      if ( Cache.enabled ) {
+
+        // Allow other file loaders to wait and subscribe on this request.
+        currentlyLoadingFiles[ url ] = [];
+
+      }
+
 			request.open( 'GET', url, true );
```
Comment (takahirox) [inline src/loaders/FileLoader.js]: Wrong indents.

File: src/loaders/FileLoader.js [hunk 3]
```diff
@@ -163,6 +186,21 @@ Object.assign( FileLoader.prototype, {
...
+          // Tell other requests for the same file that the file has finished loading.
+          for ( var i = 0; i < duplicateRequests.length; i ++ ) {
+
+            duplicateRequests[i]( response );
+
+          }
+
+        }
+
 			}, false );
```
Comment (takahirox) [inline src/loaders/FileLoader.js]: Wrong indents.

File: src/loaders/FileLoader.js [hunk 4]
```diff
@@ -39,6 +41,19 @@ Object.assign( FileLoader.prototype, {
 
 		}
 
+		// If file is already in process of loading, wait for it to load.
+		if ( currentlyLoadingFiles[ url ] ) {
+
```
Comment (takahirox) [inline src/loaders/FileLoader.js]: Personally I prefer strict comparison with `undefined` for the consistent code in `FileLoader`.

**Comment by ngokevin** [comment]
Updated.

**Comment by mrdoob** [comment]
Thanks!

**Comment by mrdoob** [comment]
Actually, I think this could be default (no need to check for `Cache.enabled`). Will change.

**Comment by takahirox** [comment]
Comment from @bananer https://github.com/mrdoob/three.js/pull/12421#issuecomment-338134918

 > in case loading the file fails, only the first request would receive the error callback

**Comment by takahirox** [comment]
And probably we need `manager.itemStart( url )` when `loading[ url ].push()` ?

**Comment by Mugen87** [comment]
I think we don't need it...

**Pull request merged**